File: sample/Julia/30_stateful_fit_predict.jl
# Objective: Stateful mini-model — fit a line, then predict; optional plot.
#. Note_1 - Arm / Ctrl+C one chunk at a time (highlighter Note1).
#.. Note_2 - blank lines are sniper chunks, not noise (highlighter Note2).
#... Note_3 - Notes never execute; they only colour the transcript (Note3).
# Progressive REPL: coefficients remain for the next clipboard chunk.
# Blank lines = sniper chunks. No blanks inside function bodies.

using Random

Random.seed!(11)

function fit_line(xs, ys)
    n = length(xs)
    mx = sum(xs) / n
    my = sum(ys) / n
    num = sum((x - mx) * (y - my) for (x, y) in zip(xs, ys))
    den = sum((x - mx)^2 for x in xs)
    slope = num / den
    intercept = my - slope * mx
    return (intercept, slope)
end

function predict(beta, xs)
    a, b = beta
    return [a + b * x for x in xs]
end

# Step A — synthetic data
xs = [i / 10 for i in 0:49]
ys = [2.5 + 1.7 * x + 0.4 * randn() for x in xs]

# Step B — fit (β survives for later chunks)
beta = fit_line(xs, ys)
println("beta=", round.(collect(beta); digits = 4))

# Step C — predict + SSE
yhat = predict(beta, xs)
sse = sum(abs2, ys .- yhat)
println("sse=", round(sse; digits = 4))
@assert sse < 20
println("PASS 30_stateful_fit_predict")

#. --- GRAPHIC OUTPUT ---

# Optional Plots.jl overlay (Plot panel / headless)
try
    ENV["GKSwstype"] = "100"
    using Plots
    scatter(xs, ys;
            ms = 8,
            color = :darkorange,
            markerstrokewidth = 1,
            markerstrokecolor = :black,
            alpha = 0.85,
            label = "data",
            title = "sample 30 — fit / predict",
            show = false)
    plot!(xs, yhat;
          lw = 3.5,
          color = :navy,
          label = "fit",
          show = false)
    println("PASS 30_plot")
catch e
    println("SKIP 30_plot (Pkg): ", e)
end

File: sample/Python/30_stateful_fit_predict.py
# Objective: Stateful mini-model — fit a line, then predict; optional plot.
# Progressive REPL: coefficients remain for the next clipboard chunk.
# No blank lines inside def bodies (progressive REPL closes on blank).

import random

random.seed(30)

def fit_line(xs, ys):
    n = len(xs)
    mx = sum(xs) / n
    my = sum(ys) / n
    num = sum((x - mx) * (y - my) for x, y in zip(xs, ys))
    den = sum((x - mx) ** 2 for x in xs)
    slope = num / den
    intercept = my - slope * mx
    return intercept, slope

def predict(beta, xs):
    a, b = beta
    return [a + b * x for x in xs]

xs = [i / 10 for i in range(50)]
ys = [2.5 + 1.7 * x + random.gauss(0, 0.4) for x in xs]
beta = fit_line(xs, ys)
print("beta=", tuple(round(v, 4) for v in beta))

yhat = predict(beta, xs)
sse = sum((y - yh) ** 2 for y, yh in zip(ys, yhat))
print("sse=", round(sse, 4))
assert sse < 20
print("PASS 30_stateful_fit_predict")

# --- GRAPHIC OUTPUT ---

try:
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt
except ImportError:
    print("SKIP 30_plot (no matplotlib)")
else:
    fig, ax = plt.subplots()
    ax.scatter(xs, ys, s=16, alpha=0.7, label="data")
    ax.plot(xs, yhat, color="crimson", linewidth=2, label="fit")
    ax.set_title("sample 30 — fit / predict")
    ax.legend()
    fig.tight_layout()
    print("PASS 30_plot")
